File: cpp/src/core/augmentations/drift/linear_drift.c
/* SPDX-License-Identifier: CECILL-2.1 */
/*
 * LinearBaselineDrift — reference implementation.
 *
 * RNG draw order matches `linear_drift.h`:
 *   1. `rows` offset uniforms in [offset_lo, offset_hi)
 *   2. `rows` slope  uniforms in [slope_lo,  slope_hi )
 *
 * The wavelength axis is taken as `arange(n_features)` (the implicit branch
 * in `nirs4all.spectral.LinearBaselineDrift._transform_impl`); the centered
 * axis is `arange - mean(arange) = arange - (n_features - 1)/2`. The mean
 * is computed exactly as `(cols - 1) / 2.0` to avoid any drift from
 * accumulating arange in a loop.
 */

#include "linear_drift.h"

#include <stdlib.h>

struct c4a_aug_linear_drift_state_t {
    c4a_rng_pcg64* rng;   /* borrowed; not owned. */
    double offset_min;
    double offset_max;
    double slope_min;
    double slope_max;
};

c4a_aug_linear_drift_state_t* c4a_aug_linear_drift_state_new(
    c4a_rng_pcg64* rng,
    double offset_min, double offset_max,
    double slope_min,  double slope_max) {
    c4a_aug_linear_drift_state_t* s =
        (c4a_aug_linear_drift_state_t*)malloc(sizeof(*s));
    if (s == NULL) return NULL;
    s->rng        = rng;
    s->offset_min = offset_min;
    s->offset_max = offset_max;
    s->slope_min  = slope_min;
    s->slope_max  = slope_max;
    return s;
}

void c4a_aug_linear_drift_state_free(c4a_aug_linear_drift_state_t* state) {
    free(state);
}

c4a_status_t c4a_aug_linear_drift_state_apply(
    c4a_aug_linear_drift_state_t* state,
    const double* X, int64_t rows, int64_t cols, double* out) {
    if (state == NULL || state->rng == NULL) return C4A_ERR_NULL_POINTER;
    if (X == NULL || out == NULL)             return C4A_ERR_NULL_POINTER;
    if (rows < 0 || cols < 0)                 return C4A_ERR_INVALID_ARGUMENT;
    if (state->offset_max < state->offset_min ||
        state->slope_max  < state->slope_min) {
        return C4A_ERR_INVALID_ARGUMENT;
    }
    if (rows == 0 || cols == 0) return C4A_OK;

    double* offsets = (double*)malloc((size_t)rows * sizeof(double));
    if (offsets == NULL) return C4A_ERR_OUT_OF_MEMORY;
    const double off_span = state->offset_max - state->offset_min;
    for (int64_t i = 0; i < rows; ++i) {
        const double u = c4a_pcg64_engine_next_double(state->rng);
        offsets[(size_t)i] = state->offset_min + off_span * u;
    }

    double* slopes = (double*)malloc((size_t)rows * sizeof(double));
    if (slopes == NULL) {
        free(offsets);
        return C4A_ERR_OUT_OF_MEMORY;
    }
    const double slope_span = state->slope_max - state->slope_min;
    for (int64_t i = 0; i < rows; ++i) {
        const double u = c4a_pcg64_engine_next_double(state->rng);
        slopes[(size_t)i] = state->slope_min + slope_span * u;
    }

    /* lambdas_centered = j - (n_features - 1)/2 */
    const double mean_j = (double)(cols - 1) * 0.5;
    for (int64_t i = 0; i < rows; ++i) {
        const double a  = offsets[(size_t)i];
        const double b  = slopes[(size_t)i];
        const size_t off = (size_t)i * (size_t)cols;
        for (int64_t j = 0; j < cols; ++j) {
            const double lc = (double)j - mean_j;
            out[off + (size_t)j] = X[off + (size_t)j] + a + b * lc;
        }
    }

    free(slopes);
    free(offsets);
    return C4A_OK;
}

File: parity/python_generator/src/c4a_parity_augmenters_ref/linear_drift.py
# SPDX-License-Identifier: CECILL-2.1
"""
LinearBaselineDrift — frozen NumPy reference.

Per-sample affine baseline drift with uniform-drawn offsets and slopes:

    offsets[i] = offset_min + (offset_max - offset_min) * u_off_i
    slopes[i]  = slope_min  + (slope_max  - slope_min)  * u_slope_i
    out[i, j]  = X[i, j] + offsets[i] + slopes[i] * (j - (n_features - 1)/2)

where ``u_off_i, u_slope_i ~ Uniform[0, 1)`` are drawn from
``rng.random(...)`` (a.k.a. ``next_double`` in the PCG64 contract).

The wavelength axis defaults to ``arange(n_features)`` and is centered;
nirs4all subtracts the mean of ``arange``, which equals ``(n - 1) / 2``.

RNG draw order (locked):
  1. n_samples uniforms for offsets
  2. n_samples uniforms for slopes
"""

from __future__ import annotations

import numpy as np


__all__ = ["linear_drift_apply"]


def linear_drift_apply(
    X: np.ndarray,
    *,
    rng: np.random.Generator,
    offset_min: float,
    offset_max: float,
    slope_min: float,
    slope_max: float,
) -> np.ndarray:
    """Apply per-sample linear baseline drift to spectra.

    Parameters
    ----------
    X : np.ndarray, shape (n_samples, n_features), dtype float64
    rng : np.random.Generator
    offset_min, offset_max : float
        Uniform range for the per-sample offset term.
    slope_min, slope_max : float
        Uniform range for the per-sample slope term.

    Returns
    -------
    out : np.ndarray, same shape and dtype as X
    """
    X = np.ascontiguousarray(X, dtype=np.float64)
    if X.ndim != 2:
        raise ValueError(f"X must be 2-D; got {X.ndim}-D")
    if offset_max < offset_min or slope_max < slope_min:
        raise ValueError("require min <= max for offsets and slopes")

    n_samples, n_features = X.shape
    if n_samples == 0 or n_features == 0:
        return X.copy()

    off_span = offset_max - offset_min
    slope_span = slope_max - slope_min

    u_offsets = rng.random(size=n_samples)
    u_slopes = rng.random(size=n_samples)
    offsets = offset_min + off_span * u_offsets
    slopes = slope_min + slope_span * u_slopes

    mean_j = (n_features - 1) * 0.5
    lambdas_centered = np.arange(n_features, dtype=np.float64) - mean_j
    drift = offsets.reshape(-1, 1) + slopes.reshape(-1, 1) * lambdas_centered.reshape(1, -1)
    return X + drift
